# 02 — Preprocessing & Imbalance Strategies

This notebook builds the shared preprocessing pipeline (scaling, 3-way split) and compares three approaches to handling class imbalance: SMOTE, class weighting, and threshold tuning.

## 1. Load and Clean Data

Reload the raw dataset and reapply the two fixes established in EDA: casting `Class` to integer and removing the 1,081 duplicate rows. This notebook is independent of `01_eda.ipynb`, so these steps are repeated here — once we move this logic into `src/features.py`, later notebooks will just import it instead of repeating it.

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

df = pd.read_csv('../data/raw/creditcard.csv')
df['Class'] = df['Class'].astype(str).str.strip("'").astype(int)
df = df.drop_duplicates()

print(df.shape)
print(df['Class'].value_counts())

(283726, 31)
Class
0    283253
1       473
Name: count, dtype: int64


## 2. Scale `Amount` and `Time`

The `V1`–`V28` columns are already PCA components (roughly zero-mean, unit-variance) and should **not** be scaled again. Only `Amount` and `Time` are raw, unscaled features — apply `StandardScaler` to just these two so they're on a comparable scale to the PCA features for models that are scale-sensitive (Logistic Regression, and later the conformal prediction calibration).

In [3]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
df[['Amount', 'Time']] = scaler.fit_transform(df[['Amount', 'Time']])

df[['Amount', 'Time']].describe()

,Amount,Time
count,2.837260e+05,2.837260e+05
mean,-8.013847e-19,1.538659e-16
std,1.000002e+00,1.000002e+00
min,-3.533268e-01,-1.996823e+00
25%,-3.309625e-01,-8.552128e-01
50%,-2.654671e-01,-2.131081e-01
75%,-4.378088e-02,9.369423e-01
max,1.022476e+02,1.642362e+00


## 3. Three-Way Stratified Split: Train / Calibration / Test

Split the data into three sets instead of the usual two. The calibration set is required later for the conformal prediction wrapper (Layer 4) — MAPIE needs data the model hasn't trained on to produce valid confidence intervals.

Split ratios: 60% train / 20% calibration / 20% test. `stratify=` is essential here — with only 473 fraud cases in the entire dataset, a non-stratified split could easily leave one of the three sets with very few (or disproportionate) fraud examples.

In [4]:
X = df.drop(columns=['Class'])
y = df['Class']

# First split: 60% train, 40% temp (which becomes calibration + test)
X_train, X_temp, y_train, y_temp = train_test_split(
    X, y, test_size=0.4, stratify=y, random_state=42
)

# Second split: split the 40% temp into 20% calibration, 20% test
X_calib, X_test, y_calib, y_test = train_test_split(
    X_temp, y_temp, test_size=0.5, stratify=y_temp, random_state=42
)

print("Train:", X_train.shape, y_train.sum(), "fraud")
print("Calibration:", X_calib.shape, y_calib.sum(), "fraud")
print("Test:", X_test.shape, y_test.sum(), "fraud")

Train: (170235, 30) 284 fraud
Calibration: (56745, 30) 94 fraud
Test: (56746, 30) 95 fraud


## 4. Save the Split Datasets

Persist the three splits to disk as Parquet files so every subsequent notebook (model comparison, SHAP, conformal prediction, drift, business impact) can load the exact same split without re-running preprocessing — ensuring consistency across the whole project.

In [5]:
train_df = X_train.copy()
train_df['Class'] = y_train

calib_df = X_calib.copy()
calib_df['Class'] = y_calib

test_df = X_test.copy()
test_df['Class'] = y_test

train_df.to_parquet('../data/processed/train.parquet', index=False)
calib_df.to_parquet('../data/processed/calibration.parquet', index=False)
test_df.to_parquet('../data/processed/test.parquet', index=False)

print("Saved train, calibration, and test sets to data/processed/")

Saved train, calibration, and test sets to data/processed/


## 5. Baseline Model — No Imbalance Handling

Before comparing imbalance-handling strategies, establish a naive baseline: train a plain Logistic Regression with no special handling of the imbalance. This sets the floor we're trying to improve on, and demonstrates concretely why naive accuracy is a misleading metric here.

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

baseline_model = LogisticRegression(max_iter=1000, random_state=42)
baseline_model.fit(X_train, y_train)

y_pred = baseline_model.predict(X_test)

print(classification_report(y_test, y_pred, digits=4))
print(confusion_matrix(y_test, y_pred))

              precision    recall  f1-score   support

           0     0.9993    0.9998    0.9996     56651
           1     0.8657    0.6105    0.7160        95

    accuracy                         0.9992     56746
   macro avg     0.9325    0.8052    0.8578     56746
weighted avg     0.9991    0.9992    0.9991     56746

[[56642     9]
 [   37    58]]


## 6. Strategy 1 — SMOTE (Synthetic Minority Oversampling)

SMOTE generates synthetic fraud examples by interpolating between existing fraud cases in feature space, rather than simply duplicating them. Applied only to the **training set** — never to calibration or test — to avoid leaking synthetic patterns into evaluation data.

In [7]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)
X_train_smote, y_train_smote = smote.fit_resample(X_train, y_train)

print("Before SMOTE:", y_train.value_counts().to_dict())
print("After SMOTE:", y_train_smote.value_counts().to_dict())

smote_model = LogisticRegression(max_iter=1000, random_state=42)
smote_model.fit(X_train_smote, y_train_smote)

y_pred_smote = smote_model.predict(X_test)
print(classification_report(y_test, y_pred_smote, digits=4))
print(confusion_matrix(y_test, y_pred_smote))

Before SMOTE: {0: 169951, 1: 284}
After SMOTE: {0: 169951, 1: 169951}
              precision    recall  f1-score   support

           0     0.9998    0.9752    0.9874     56651
           1     0.0571    0.8947    0.1074        95

    accuracy                         0.9751     56746
   macro avg     0.5285    0.9350    0.5474     56746
weighted avg     0.9982    0.9751    0.9859     56746

[[55248  1403]
 [   10    85]]


## 7. Strategy 2 — Class Weighting

Instead of creating synthetic samples, `class_weight='balanced'` tells the model to penalize misclassifying the minority class (fraud) more heavily during training — no synthetic data, no change to the training set size.

In [8]:
weighted_model = LogisticRegression(max_iter=1000, class_weight='balanced', random_state=42)
weighted_model.fit(X_train, y_train)

y_pred_weighted = weighted_model.predict(X_test)
print(classification_report(y_test, y_pred_weighted, digits=4))
print(confusion_matrix(y_test, y_pred_weighted))

              precision    recall  f1-score   support

           0     0.9998    0.9770    0.9883     56651
           1     0.0619    0.9053    0.1158        95

    accuracy                         0.9769     56746
   macro avg     0.5309    0.9411    0.5521     56746
weighted avg     0.9983    0.9769    0.9868     56746

[[55347  1304]
 [    9    86]]


## 8. Strategy 3 — Threshold Tuning

Rather than modifying the training data or the loss function, train a plain model (like the baseline) and instead adjust the **decision threshold** at inference time — the probability cutoff above which a transaction is classified as fraud. Default is 0.5; lowering it increases recall at the cost of precision, and vice versa. This gives explicit, tunable control over the precision-recall trade-off, which is the foundation for the business-impact threshold explorer in later layers.
